File: java-version/src/main/java/bernard/tatin/threads/ProtectedFlag.java
package bernard.tatin.threads;

public class ProtectedFlag {
    private boolean flag = false;
    private final Mutex mutex = new Mutex();

    public ProtectedFlag() {

    }

    public ProtectedFlag(boolean initialValue) {
        flag = initialValue;
    }

    private void setV(boolean newValue) {
        try {
            mutex.lock();
        } catch (InterruptedException e) {

        }
        flag = newValue;
        mutex.unlock();
    }

    public void set() {
        setV(true);
    }
    public void reset() {
        setV(false);
    }
    public boolean get() {
        try {
            mutex.lock();
        } catch (InterruptedException e) {

        }
        boolean r = flag;
        mutex.unlock();
        return r;
    }

}

File: kotlin-version/src/main/kotlin/bernard/tatin/Threads/ProtectedFlag.kt
package bernard.tatin.Threads

class ProtectedFlag {
    private var flag = false
    private val mutex = Mutex()

    constructor() {

    }

    constructor(initialValue: Boolean) {
        flag = initialValue
    }

    private fun setV(newValue: Boolean) {
        try {
            mutex.lock()
        } catch (e: InterruptedException) {

        }

        flag = newValue
        mutex.unlock()
    }

    fun set() {
        setV(true)
    }

    fun reset() {
        setV(false)
    }

    fun get(): Boolean {
        try {
            mutex.lock()
        } catch (e: InterruptedException) {

        }

        val r = flag
        mutex.unlock()
        return r
    }

}
